From 0a8d04eeac95f4db9d03ee31070bd825a7feb0b2 Mon Sep 17 00:00:00 2001
From: Matthieu Herrb <matthieu.herrb@laas.fr>
Date: Sun, 01 Jan 2012 17:27:54 +0000
Subject: Update for new vgaHW API.

Signed-off-by: Matthieu Herrb <matthieu.herrb@laas.fr>
Reviewed-by: Mark Kettenis <kettenis@openbsd.org>
Signed-off-by: Alex Deucher <alexdeucher@gmail.com>
---
diff --git a/src/radeon_driver.c b/src/radeon_driver.c
index a9abd66..81f5645 100644
--- a/src/radeon_driver.c
+++ b/src/radeon_driver.c
@@ -3131,8 +3131,10 @@ Bool RADEONPreInit(ScrnInfoPtr pScrn, int flags)
     } else
            xf86DrvMsg(pScrn->scrnIndex, X_INFO, "VGAAccess option set to FALSE,"
                       " VGA module load skipped\n");
-    if (info->VGAAccess)
+    if (info->VGAAccess) {
+	vgaHWSetStdFuncs(VGAHWPTR(pScrn));
         vgaHWGetIOBase(VGAHWPTR(pScrn));
+    }
 #endif
 
 
--
cgit v0.9.0.2-2-gbebe
