commit 3f9b597dedc45379b0bc0b631f3f924c403bca48
Author: Dave Airlie <airlied@linux.ie>
Date:   Fri Dec 21 09:55:42 2007 +1000

    fixup clip test include handling

diff --git a/configure.ac b/configure.ac
index 91ea4a8..159eced 100644
--- a/configure.ac
+++ b/configure.ac
@@ -221,7 +221,10 @@ AC_CHECK_DECL(xf86ModeBandwidth,
 AC_CHECK_DECL(xf86_crtc_clip_video_helper,
 	      [AC_DEFINE(HAVE_XF86CRTCCLIPVIDEOHELPER, 1, [Have xf86_crtc_clip_video_helper prototype])],
 	      [],
-	      [#include "xf86Crtc.h"])
+	      [#include <X11/Xmd.h>
+	       #include "xorg-server.h"
+	       #include "xf86i2c.h"
+	       #include "xf86Crtc.h"])
 
 AC_CHECK_DECL(XSERVER_LIBPCIACCESS,
 	      [XSERVER_LIBPCIACCESS=yes],[XSERVER_LIBPCIACCESS=no],
commit be7f8fd338f5af8b632f16a83db41e15d00af469
Author: Dave Airlie <airlied@linux.ie>
Date:   Fri Dec 21 10:03:49 2007 +1000

    fix mode bandwidth configure check

diff --git a/configure.ac b/configure.ac
index 159eced..edac3ed 100644
--- a/configure.ac
+++ b/configure.ac
@@ -216,7 +216,8 @@ AC_CHECK_DECL(xf86XVFillKeyHelperDrawable,
 AC_CHECK_DECL(xf86ModeBandwidth,
 	      [AC_DEFINE(HAVE_XF86MODEBANDWIDTH, 1, [Have xf86ModeBandwidth prototype])],
 	      [],
-	      [#include "xf86Modes.h"])
+	      [#include "xorg-server.h"
+	       #include "xf86Modes.h"])
 
 AC_CHECK_DECL(xf86_crtc_clip_video_helper,
 	      [AC_DEFINE(HAVE_XF86CRTCCLIPVIDEOHELPER, 1, [Have xf86_crtc_clip_video_helper prototype])],
